theano/tensor/nlinalg.py

```diff
@@ -292,7 +289,7 @@ def grad(self, inputs, g_outputs):
         return [gz * self(x) * matrix_inverse(x).T]
 
     def infer_shape(self, node, shapes):
-        return [()]
+        return [shapes[0]]
```

Comment:
In the old case when x.ndim == 2, this don't return the right answer.

What could be done:
```
if node.inputs[0].ndim == 2:
   return [()]
raise NotImplementedError()
```

This would make infer_shape not implemented for other ndim. I don't know what is det(nd tensor), so maybe something more generic can be done.